src/ast/proof.rs

```diff
@@ -0,0 +1,556 @@
+//! Parsing for proof statements
+
+use crate::ast::{Error, ErrorContext, ErrorKind, Ident, ParseRet};
+use crate::tokens::{self, Oper, Token, TokenKind};
+use std::convert::{TryFrom, TryInto};
+
+/// Consumes all given lines of proof
+pub(super) fn consume_proof_lines<'a>(
+    tokens: &'a [Token<'a>],
+) -> (usize, ParseRet<'a, Vec<Stmt<'a>>>) {
+    let mut stmts = Vec::new();
+    let mut errors = Vec::new();
+
+    for (consumed, t) in tokens.iter().enumerate() {
+        let ts = match &t.kind {
+            TokenKind::ProofLine(ts) => ts,
+            _ => return (consumed, ParseRet::with_soft_errs(stmts, errors)),
+        };
+
+        // The success type of `Stmt::parse` is an `Option<Stmt>`, because empty proof lines are
+        // allowed.
+        let stmt: Option<Stmt> = match Stmt::parse(ts) {
+            ParseRet::Ok(s) => s,
+            ParseRet::SoftErr(s, es) => {
+                errors.extend(es);
+                s
+            }
+            ParseRet::Err(es) => {
+                errors.extend(es);
+                return (consumed, ParseRet::Err(errors));
+            }
+        };
+
+        if let Some(s) = stmt {
+            stmts.push(s);
+        }
+    }
+
+    (tokens.len(), ParseRet::with_soft_errs(stmts, errors))
+}
+
+/// A single proof statement
+///
+/// While this shares a name with `ast::Stmt`, it bears no other relation to that type.
+#[derive(Debug, Clone)]
+pub struct Stmt<'a> {
+    pub kind: StmtKind<'a>,
+
+    /// The source of the proof statement. These will always exactly use a single line, so the
+    /// source token has kind `TokenKind::ProofLine`.
+    source: &'a [Token<'a>],
+}
+
+#[derive(Debug, Clone)]
+pub enum StmtKind<'a> {
+    /// A 'require' statement indicates some form of strict requirement for function arguments.
+    /// Here are a couple examples:
+    /// ```text
+    /// # require x <= 4
+    /// ```
+    /// and
+    /// ```text
+    /// # require x + y >= z - x
+    /// ```
+    Require(Condition<'a>),
+
+    /// A contract indicates a promise by the function; if a certain condition is met of the
+    /// inputs, a separate condition will be met by the output.
+    ///
+    /// Conditions in contracts are separated by the 'implies' operator, `=>`. Here is an example:
+    /// ```text
+    /// # x < 4 => _ < 3
+    /// ```
+    /// Note that the return value is referred to by the identifier `_`.
+    Contract {
+        /// The pre-condition
+        ///
+        /// This is optional in order to allow unconditional return values. This could be written
+        /// as the following:
+        /// ```text
+        /// # => _ >= x
```

Comment:
Implies is the right symbol to use here however `=>` doesn't jump out when used with `<=` and `>=`.
I think it's the type of syntax which is important when reading and so should stand out.
`->` may be a sensible symbol too?
Implies is probably the more correct though. How about `==>`?
I don't have strong feelings though.

Reply:
That all makes sense. It should be fairly trivial to change if we come up with something better, so maybe we do the classic "let's stick the mediocre thing *for now* and pretend that we'll come back and change it"